### PR TITLE
feat(web): allows shifting the recognition zone of an active GestureSource 🐵

### PR DESCRIPTION
**Pull request by jahorton**
This PR allows "pushing" an alternate configuration for the engine's "recognition zone" for an ongoing gesture.  This is particularly designed to model the shift from a longpress to subkey-selection, which is generally within a smaller and more restrictive area.

The need for this was discovered as part of #9525; the changes are easily significant enough to merit spinning it off in this manner, as it'll ease that future review.

It also includes a number of new unit tests for `GestureSource` and adds new "guards" to certain methods, prohibiting a subview from performing update commands either on or separate from its underlying, base Source.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts [hunk 1]
```diff
@@ -81,4 +84,34 @@ export interface GestureRecognizerConfiguration<HoveredItemType> {
    * @returns
    */
   readonly itemIdentifier?: (coord: Omit<InputSample<any>, 'item'>, target: EventTarget) => HoveredItemType;
+}
+
+export function preprocessRecognizerConfig<HoveredItemType>(
+  config: GestureRecognizerConfiguration<HoveredItemType>
+): Nonoptional<GestureRecognizerConfiguration<HoveredItemType>> {
+  // Allows configuration pre-processing during this method.
+  let processingConfig: Mutable<Nonoptional<GestureRecognizerConfiguration<HoveredItemType>>> = {...config} as Nonoptional<GestureRecognizerConfiguration<HoveredItemType>>;
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:93]: That is one heck of a type -- starting to feel downright Java!

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts [hunk 2]
```diff
@@ -81,4 +84,34 @@ export interface GestureRecognizerConfiguration<HoveredItemType> {
...
+  processingConfig.mouseEventRoot = processingConfig.mouseEventRoot ?? processingConfig.targetRoot;
+  processingConfig.touchEventRoot = processingConfig.touchEventRoot ?? processingConfig.targetRoot;
+
+  processingConfig.inputStartBounds = processingConfig.inputStartBounds ?? processingConfig.targetRoot;
+  processingConfig.maxRoamingBounds = processingConfig.maxRoamingBounds ?? processingConfig.targetRoot;
+  processingConfig.safeBounds       = processingConfig.safeBounds       ?? new PaddedZoneSource([2]);
+
+  processingConfig.itemIdentifier   = processingConfig.itemIdentifier   ?? (() => null);
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:101]: Some comments on why these are being setup this way would be helpful

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts [hunk 3]
```diff
@@ -81,4 +84,34 @@ export interface GestureRecognizerConfiguration<HoveredItemType> {
...
+
+  if(!config.paddedSafeBounds) {
+    let paddingArray = config.safeBoundPadding;
+    if(typeof paddingArray == 'number') {
+      paddingArray = [ paddingArray ];
+    }
+    paddingArray = paddingArray ?? [3];
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:108]: Why 3? Magic?

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts [hunk 4]
```diff
@@ -81,4 +84,34 @@ export interface GestureRecognizerConfiguration<HoveredItemType> {
...
+  processingConfig.touchEventRoot = processingConfig.touchEventRoot ?? processingConfig.targetRoot;
+
+  processingConfig.inputStartBounds = processingConfig.inputStartBounds ?? processingConfig.targetRoot;
+  processingConfig.maxRoamingBounds = processingConfig.maxRoamingBounds ?? processingConfig.targetRoot;
+  processingConfig.safeBounds       = processingConfig.safeBounds       ?? new PaddedZoneSource([2]);
+
+  processingConfig.itemIdentifier   = processingConfig.itemIdentifier   ?? (() => null);
+
+  if(!config.paddedSafeBounds) {
+    let paddingArray = config.safeBoundPadding;
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:104]: These two member names are super conflatable!

File: common/web/gesture-recognizer/src/engine/gestureRecognizer.ts [hunk 5]
```diff
@@ -12,39 +10,9 @@ export class GestureRecognizer<HoveredItemType> extends TouchpointCoordinator<Ho
   private readonly mouseEngine: MouseEventEngine<HoveredItemType>;
   private readonly touchEngine: TouchEventEngine<HoveredItemType>;
 
-  protected static preprocessConfig<HoveredItemType>(
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/gestureRecognizer.ts:15]: And ... now I see that my comments relate to pre-existing code. I still think they are good comments though :grin:

File: common/web/gesture-recognizer/src/engine/headless/gestureSource.ts [hunk 6]
```diff
@@ -167,25 +203,61 @@ export class GestureSourceSubview<HoveredItemType> extends GestureSource<Hovered
...
+    configStack: typeof GestureSource.prototype['recognizerConfigStack'],
+    startAtEnd: boolean,
+    preserveBaseItem: boolean
+  ) {
+    let mayUpdate = true;
+    let start = 0;
+    let length = source.path.coords.length;
+    if(source instanceof GestureSourceSubview) {
+      const expectedLength = source._baseStartIndex + source.path.coords.length;
+      start = source._baseStartIndex;
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/headless/gestureSource.ts:217]: Given we've just declared it above, I think this is also clearer being in the same order:

```suggestion
      start = source._baseStartIndex;
      const expectedLength = start + length;
```

File: common/web/gesture-recognizer/src/engine/headless/gestureSource.ts [hunk 7]
```diff
@@ -167,25 +203,61 @@ export class GestureSourceSubview<HoveredItemType> extends GestureSource<Hovered
...
+      length = length > 0 ? 1 : 0;
     } else {
-      subpath = source.path.clone();
+      this._baseStartIndex = start;
+    }
+
+    subpath = new GesturePath<HoveredItemType>();
+    for(let i=0; i < length; i++) {
+      const index = start + i;
+      subpath.extend(translateSample(baseSource.path.coords[index]));
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/headless/gestureSource.ts:260]:
```suggestion
      subpath.extend(translateSample(baseSource.path.coords[start + i]));
```

Not sure the additional variable gives us much here :grin:

File: common/web/gesture-recognizer/src/test/resources/simulateMultiSourceInput.ts [hunk 8]
```diff
@@ -33,7 +33,7 @@ interface MockedPredecessor<Type> {
     sample: InputSample<Type>,
     baseItem: Type
   },
-  sources: GestureSource<Type>[],
+  sources: GestureSource<Type>[], // FIX ME.
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/test/resources/simulateMultiSourceInput.ts:36]: // FIX ME

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Looking good, I think. I am struggling to keep the abstractions in my head and relate them to how gestures work though!

I see that the build failed, but here's some review feedback anyway 😀

**Comment by jahorton** [comment]
Accidentally pushed a new commit to this PR instead of the one after it in the chain; the force-push simply removed that one commit.

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts (hunk 1 above)
Comment (jahorton) [reply common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:93]: I always have loved generics a little _too_ much.

Obviously not _that_ extreme, but it's been noted that [TypeScript's type system is Turing-complete](https://github.com/microsoft/TypeScript/issues/14833).

Okay, maybe I don't love them _that_ much - some people clearly had fun writing their analyses in that issue.  (Like the guy who wrote a MUD with them!)

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts (hunk 2 above)
Comment (jahorton) [reply common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:101]: Note that the documentation for the fields and their defaults has already been established in JSDoc-style comments on each of them.  It shows up via IntelliSense within VS Code.

Granted, _why_ the defaults are set as they are... that is a different question.  Is _that_ your concern?

Edit:

> And ... now I see[...]

Ah.  Got it, the concern would make more sense if the code were appearing without any context.

File: common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts (hunk 3 above)
Comment (jahorton) [reply common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts:108]: https://github.com/keymanapp/keyman/blob/be196744fd1fef9baf138072b2a5e44c95da6d7b/common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts#L58-L59

From:

https://github.com/keymanapp/keyman/blob/be196744fd1fef9baf138072b2a5e44c95da6d7b/common/web/gesture-recognizer/src/engine/configuration/gestureRecognizerConfiguration.ts#L52-L63

If memory serves, this default is derived from the corresponding value KMW is already using for the OSK.  I've just abstracted its purpose and given it a name and documentation.

**Review by mcdurdin — APPROVED** [review]

**Pull request merged**